pkg/stub/handler.go

```diff
@@ -16,75 +19,158 @@ import (
 )
 
 func NewHandler() sdk.Handler {
+	operatorNamespace, err := k8sutil.GetWatchNamespace()
+	if err != nil {
+		logrus.Fatalf("Failed to get watch namespace: %v", err)
+	}
+
+	operatorName, err := k8sutil.GetOperatorName()
+	if err != nil {
+		logrus.Fatalf("Failed to get operator name: %v", err)
+	}
+
 	return &Handler{
-		manifestFactory: manifests.NewFactory(),
+		manifestFactory:   manifests.NewFactory(),
+		operatorNamespace: operatorNamespace,
+		operatorName:      operatorName,
 	}
 }
 
 type Handler struct {
 	manifestFactory *manifests.Factory
+
+	operatorNamespace string
+	operatorName      string
 }
 
 func (h *Handler) Handle(ctx context.Context, event sdk.Event) error {
 	switch o := event.Object.(type) {
 	case *ingressv1alpha1.ClusterIngress:
 		if event.Deleted {
+			cond := osv1.OperatorStatusCondition{
+				Type:    osv1.OperatorStatusConditionTypeWorking,
+				Message: fmt.Sprintf("working towards state: ClusterIngress %q deleted",
+					o.Name),
+			}
+			if err := h.syncOperatorStatus(cond); err != nil {
+				return err
+			}
+
 			logrus.Infof("Deleting ClusterIngress object: %s", o.Name)
-			return h.deleteIngress(o)
-		} else {
-			return h.syncIngressUpdate(o)
+			if err := h.deleteIngress(o); err != nil {
+				return err
+			}
+
+			cond = osv1.OperatorStatusCondition{
+				Type:    osv1.OperatorStatusConditionTypeDone,
+				Message: fmt.Sprintf("done applying state: ClusterIngress %q deleted",
+					o.Name),
+			}
+			if err := h.syncOperatorStatus(cond); err != nil {
+				return err
+			}
+
+			return nil
+		}
+
+		cond := osv1.OperatorStatusCondition{
+			Type:    osv1.OperatorStatusConditionTypeWorking,
+			Message: fmt.Sprintf("working towards state: %#v", o),
 		}
+		if err := h.syncOperatorStatus(cond); err != nil {
+			return err
+		}
+
+		if err := h.syncIngressUpdate(o); err != nil {
```

Comment:
The `syncOperatorStatus` calls woven throughout the update logic seem more imperative than level driven to me... would it be possible to simply it to "if `syncIngressUpdate` returns an error, set a degraded status"? Or take the error into account when doing a state-observation driven computation unconditionally after the update, like:

1. Attempt update
2. Compute status by observing the state of the cluster after the update
    1. Take any error from (1) into account when computing in (2)

The point being it should be possible to re-compute status outside the context of the update function. That is, there could be a `Handler.ComputeOperatorStatus()` function which could be called at any time to compute the correct status absent any event to observe.

This also closes the current gap with status computation based on deletes: the current code will only handle deletes in response to a delete event (something which needs refactored); if any delete processing (or status processing during delete processing) fails, it should be possible to call a general status recompilation function later on (outside the context of a delete event) and get the right status.